### ISSUE TITLE
Pretrain Roberta with another language

### ISSUE THREAD

**Issue by qwer4107** [facebookresearch/fairseq#4304]
Hello, I'm trying to pretrain Roberta with korean text. 
I have my own tokenizer(, which contains vocab / encoder).

With 'em, I followed the pretraining guide and I'm now training that.
![스크린샷 2022-03-24 오후 2 32 50](https://user-images.githubusercontent.com/71030815/159849143-d90f4c17-a0ae-4756-9aca-bc0592a11d90.png)

But when I load the model and encode some tokens(Korean one),
there's unknown tokens, while English tokens are not...

So, is there any other way to train another language? or is not possible?

**Comment by gmryu** [comment]
- Is this `roberta` a tokenizer? Though fairseq never utilize such things.
Fairseq's RoBERTa model do not have encode and decode. 
So I kind of confused what you want to tell here.
If it is not me being foolish and you have questions about how fairseq deal with raw text data, see below.

- Fairseq-train never cares about what language you are using. It only cares those numeric token values (those id tensors). 
"Train another language" is just "train with a different tensors" and any tensors have the same embed dimension will allow the training.

----

First, you have _korean space-seperated text data_ : (the following is just an example. I do not speak korean)
```
나 는 고 양 이 를 좋 아 한 다 # I like cats. 
나는 고양이를 좋아한다 # difference is " "(space)'s positions.
```
and _a korean dict.txt_  : dict.txt looks like this
```
나 0
는 0
...
나는 0
개를 0
```
Each line is \<token\> \<frequency score\>, though socre doesn't matter because all tokens are obviously seperated by a space.

Then, if you do `fairseq-preprocess {DATA} --src-dict {that korean dict.txt} ... `, in the above example you get
`나 는 고 양 이 를 좋 아 한 다 = tensor([4,5,6,7,8,9,10,11,12,13]) # 9 spaces, so 10 tokens. 4=나, 5=는,...`
`나는 개를 좋아한다  = tensor([14, 15, 3]) # 2 spaces, so 3 tokens. The last 3 stands for <unk> because 좋아한다 do not appear in the dict.txt`

When you finished `fairseq-preproce`, the log also tells you how many tokens are replaced by `<unk>`.
(If you do not want `<unk>`, search for BBPE or byte level bpe)
Finally you use this .bin in `fairseq-train/generate/interactive`.

**Comment by qwer4107** [comment]
@gmryu 

First of all, I sincerely thank you for your comments.

'm not really familiar with FairSeq API, but as you said I trained my own sentencepiece(BPE) tokenizer, and use it as a srcdict when building 'dict.txt', which is exactly same as what you've shown..

Also, there were no many unk replacement when I saw encoding log. 

When I saw this repo, 
https://github.com/pytorch/fairseq/tree/main/examples/roberta
![스크린샷 2022-03-30 오후 3 55 31](https://user-images.githubusercontent.com/71030815/160770064-aaeb903d-8aba-4da9-83c4-fc968a5babbf.png)
They used roberta model as a tokenizer, too (It's True... this embarrassed me)!!

Well, I know how to convert fairseq roberta to Huggingface Form, so actually there's no problem when I port the model and load my own tokenizer when using pretrained model.

But I'm worried about Pre-training Procedure whether it's training well...(When I look the log, there's no problem with gradient or loss.. but problems with 'encode'!!)

So you mean that this kind of 'encode' is non of business when pretraining? 

Thanks
qwer4107

**Comment by gmryu** [comment]
Apologize for my misunderstanding of `roberta.encode`.
I believe encode() do nothing with fairseq-train(=training and pretraining)

----
1. The point is `RobertaModel.from_pretrained('/path/to/roberta.large', checkpoint_file='model.pt')` does not return a `RobertaModel`, it returns a `GeneratorHubInterface` with a model(roberta), bpe, tokenizer inside it. The class is in `fairseq.hub_utils` 
You can see encode and decode methods here: https://github.com/pytorch/fairseq/blob/06c65c8297396959c025bf61aaeaa6f4ace9940e/fairseq/hub_utils.py#L234

So, this pops another question: If your bbpe and dict.txt have only byte-expressions, then due to the consistency -- `input x==bpe.decode(bpe.encode(x))` -- you should never get \<unk\>.  So maybe it's the tokenzier doing something wrong here? 
Also you can verify these by editting your `GeneratorHubInterface` in your cloned fairseq, add some prints, comment out some codes,etc.

2. I do not involve in fairseq's team, but from word "Generator" you can guess they are used by inference and an entropy loss do not call inference_step nor use actual words to compute scores, thus its bpe,tokenizer are never used. You can also check` fairseq_cli/train.py` and involving classes to confirm this.

**Comment by qwer4107** [comment]
@gmryu

I perfectly understand what you meant in (1) and (2) and I finally solve my problem

This is a Wrapper for model and tokenizer...etc

When I load the model and tokenizer, the inner code leads to load Hub's GPT2 Vocab(BPE based), so that there was a mismatch between my model and tokenizing methods. 

By changing some arguments, I could totally load my local tokenizing model and could encode Korean sentence well. 

Thank you for your Guideline, though my question was such an idiot.
It was really helpful for me to solve this one!

Close the issue!

**Comment by zareami10** [comment]
> @gmryu
> 
> I perfectly understand what you meant in (1) and (2) and I finally solve my problem
> 
> This is a Wrapper for model and tokenizer...etc
> 
> When I load the model and tokenizer, the inner code leads to load Hub's GPT2 Vocab(BPE based), so that there was a mismatch between my model and tokenizing methods.
> 
> By changing some arguments, I could totally load my local tokenizing model and could encode Korean sentence well.
> 
> Thank you for your Guideline, though my question was such an idiot. It was really helpful for me to solve this one!
> 
> Close the issue!

Could you please explain a bit more about how you solved the issue?
